### ISSUE TITLE
Measure Memory Overhead / Performance

### ISSUE THREAD

**Issue by benngarcia** [benngarcia/rails_async_methods#1]
Need to check out impact on performance. Already have some major optimizations in mind, though whether need to do them or not depends on how performance is impacted.

**Comment by brandonzylstra** [comment]
do we have any reason to be concerned about performance?

**Comment by benngarcia** [comment]
tl;dr - No.

My initial concern was based off [Sidekiq's performance notes regarding ActiveJob](https://github.com/sidekiq/sidekiq/wiki/Active-Job#performance).

A misstep in logic, as you're locked into ActiveJob anyway, so there is no real concern wrt job overhead. I was also concerned about overhead from parsing ruby back into ruby to keep method signatures & thrown errors accurate affecting boot times & runtime memory usage (i believe there is a better way).

I did a quick profiling assuming over 20k methods being defined via the `async` class method & being defined via `define_method` which isn't particularly useful as you'd want to compare to defining a method via `def`, but it's helpful to see what overhead is caused from specifically this gem. I provided the scripts as I vaguely feel as though I made a misstep or 2 wrt accuracy. Results:

Benchmark (would affect boot times, negligible):
       user     system      total        real
no_async:  0.069566   0.005313   0.074879 (  0.075946)
async:  1.033635   0.049272   1.082907 (  1.089120)

Memory: 
No Async:
Total allocated: 16245501 bytes (260485 objects)
Total retained:  6427160 bytes (80213 objects)


Async:
Total allocated: 69155560 bytes (1020000 objects)
Total retained:  28446670 bytes (340000 objects)

Not great. But if you have 20,000 async model methods, I'll optimize the gem for you. Different folks have different opinions, but I personally don't believe a ~1mb/1k async method memory runtime usage is something to sweat over for a rails app, nor is an additional 1 second/20k async methods added to boot times.


Leaving open as optimizing can be fun and I may reference this later.

```
require_relative './config/environment'
require 'benchmark'

Benchmark.bm do |x|
  x.report("no_async:") do
    20_000.times do |i|
      AsyncExample.define_method("instance_method_#{i}") {|a, b| 'Hello, world!' }
      AsyncExample.define_method("async_instance_method_#{i}") { |a, b| 'Hello, world!' }
    end
  end

  GC.start

  x.report("async:") do
    20_000.times do |i|
      AsyncExample.define_method("instance_method__#{i}") { |a, b| 'Hello, world!' }
      AsyncExample.async("instance_method__#{i}") { |a, b| 'Hello, world!' }
    end
  end
end
```
```
require_relative './config/environment'
require 'memory_profiler'

GC.start
report1 = MemoryProfiler.report do
  20_000.times do |i|
    AsyncExample.define_method("instance_method_#{i}") {|a, b| 'Hello, world!' }
    AsyncExample.define_method("async_instance_method_#{i}") { |a, b| 'Hello, world!' }
  end
end

report1.pretty_print(to_file: 'report1.txt')

GC.start
report2 = MemoryProfiler.report do
  20_000.times do |i|
    AsyncExample.define_method("instance_method__#{i}") { |a, b| 'Hello, world!' }
    AsyncExample.async("instance_method__#{i}") { |a, b| 'Hello, world!' }
  end
end

report2.pretty_print(to_file: 'report2.txt')
```
